service/history/api/startworkflow/api.go

```diff
@@ -359,58 +363,45 @@ func (s *Starter) resolveDuplicateWorkflowID(
 	if err != nil {
 		return nil, err
 	}
-
-	if prevExecutionUpdateAction == nil {
+	if currentExecutionUpdateAction == nil {
 		return nil, nil
 	}
-	var mutableStateInfo *mutableStateInfo
-	// update prev execution and create new execution in one transaction
-	// we already validated that currentWorkflowConditionFailed.RunID is not empty,
-	// so the following update won't try to lock current execution again.
+
+	// Update current execution and create new execution in one transaction.
+	newExecutionWorkflowLease := creationParams.workflowLease
+	currentExecutionWorkflowKey := definition.NewWorkflowKey(
+		s.namespace.ID().String(), workflowID, currentWorkflowConditionFailed.RunID)
 	err = api.GetAndUpdateWorkflowWithNew(
 		ctx,
 		nil,
 		api.BypassMutableStateConsistencyPredicate,
-		definition.NewWorkflowKey(
-			s.namespace.ID().String(),
-			workflowID,
-			currentWorkflowConditionFailed.RunID,
-		),
-		prevExecutionUpdateAction,
+		currentExecutionWorkflowKey,
+		currentExecutionUpdateAction,
 		func() (workflow.Context, workflow.MutableState, error) {
-			workflowLease, err := api.NewWorkflowWithSignal(
```

Comment:
interesting, not sure why we create a new mutable state before.

Reply:
hmmm, I think it's because the transaction for newExecutionWorkflowLease has already been closed once above. And GetAndUpdateWorkflowWithNew will try to close it again. If it works, it works in a kinda weird way.

Reply:
Oh, I see. I didn't realize that wasn't idempotent. The reason I made the change is that it's competing for the same lock we already grabbed (as part of the cache write). 🤔